ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/RAGTool.java

```diff
@@ -0,0 +1,315 @@
+    public static final String TYPE = "RAGTool";
+    private static String DEFAULT_DESCRIPTION = "Use this tool to run any model.";
+    public static final String MODEL_ID_FIELD = "model_id";
+    public static final String EMBEDDING_FIELD = "embedding_field";
+    public static final String INDEX_FIELD = "index";
+    private String name = TYPE;
+    private String description = DEFAULT_DESCRIPTION;
+    private Client client;
+    private String modelId;
+
+    private NamedXContentRegistry xContentRegistry;
+    private String index;
+    private String embeddingField;
+    private String[] sourceFields;
+    private String embeddingModelId;
+    private Integer docSize;
+    private Integer k;
+    @Setter
+    private Parser inputParser;
+    @Setter
+    private Parser outputParser;
+
+    @Builder
+    public RAGTool(
+        Client client,
+        NamedXContentRegistry xContentRegistry,
+        String index,
+        String embeddingField,
+        String[] sourceFields,
+        Integer k,
+        Integer docSize,
+        String embeddingModelId,
+        String modelId
+    ) {
+        super(client, xContentRegistry, index, sourceFields, docSize);
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+        this.index = index;
+        this.embeddingField = embeddingField;
+        this.sourceFields = sourceFields;
+        this.embeddingModelId = embeddingModelId;
+        this.docSize = docSize == null ? 2 : docSize;
+        this.k = k == null ? 10 : k;
+        this.modelId = modelId;
+
+        outputParser = new Parser() {
+            @Override
+            public Object parse(Object o) {
+                List<ModelTensors> mlModelOutputs = (List<ModelTensors>) o;
+                return mlModelOutputs.get(0).getMlModelTensors().get(0).getDataAsMap().get("response");
+            }
+        };
+    }
+
+    @Override
+    protected String getQueryBody(String queryText) {
+        if (StringUtils.isBlank(embeddingField) || StringUtils.isBlank(modelId)) {
+            throw new IllegalArgumentException(
+                "Parameter [" + EMBEDDING_FIELD + "] and [" + MODEL_ID_FIELD + "] can not be null or empty."
+            );
+        }
+        return "{\"query\":{\"neural\":{\""
+            + embeddingField
+            + "\":{\"query_text\":\""
+            + queryText
+            + "\",\"model_id\":\""
+            + modelId
+            + "\",\"k\":"
+            + k
+            + "}}}"
+            + " }";
+    }
+
+    @Override
+    public <T> void run(Map<String, String> parameters, ActionListener<T> listener) {
+        try {
+            String question = parameters.get("input");
+            try {
+                question = gson.fromJson(question, String.class);
+            } catch (Exception e) {
+                // throw new IllegalArgumentException("wrong input");
+            }
+            String query = "{\"query\":{\"neural\":{\""
```

Comment:
can't we re-use`getQueryBody` here rather than creating same query again?

Reply:
used getQueryBody in the new commit [Add static variable and reuse getQueryBody](https://github.com/opensearch-project/ml-commons/pull/1747/commits/50365af9d9220b96f868a1bce74a1db17594e9c9)